### PR TITLE
S1.0 User can publish his/her real-time location to all Ubimaps users

### PR DESCRIPTION
**Pull request by mriekkin**
Implement public location sharing:

* The backend consists of a `/public` endpoint, which has methods for fetching, adding and removing locations.
* The frontend consists of a "Share location for all users" modal. In this modal, toggling a switch enables or disables public location sharing.
* Each location is assigned a nickname, which is generated when the user enables public location sharing.
* Markers for publicly shared locations are displayed on the map, along with generated nicknames. The map shows only public location markers. All other markers are hidden.

### CODE REVIEW

File: client/src/router.tsx [hunk 1]
```diff
@@ -124,8 +131,18 @@ const Router = ({ appConfig }: Props) => {
...
   useEffect(() => {
-    const adminUser = AdminTokenStore.get();
-    setAdmin(adminUser);
+    const adminToken = AdminTokenStore.get();
+    setAdmin(adminToken);
+
+    // TODO: read also beacon token from local store
+
+    const fetchPublicBeacons = async () => {
+      const pubBeacons = await ShareLocationApi.fetchPublicBeacons();
```
Comment (mriekkin) [inline client/src/router.tsx]: Currently the tests throw an UnhandledPromiseRejectionWarning with the message "connect ECONNREFUSED 127.0.0.1:3001".

This problem disappears if I wrap the contents of fetchPublicBeacons inside a try...catch. I wonder, though, if there's some underlying cause which merits attention?

File: auth-server/src/controllers/public.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  return hri.random();
+};
+
+publicRouter.get('/', (request: Request, response: Response) => {
+  response.status(200).send(publicBeacons);
+});
+
+publicRouter.post(
+  '/',
+  (request: Request & DecodedToken, response: Response) => {
```
Comment (je-l) [inline auth-server/src/controllers/public.ts]: Please use async/await syntax here. Supertest silently ignores exceptions if async/await is not used, which makes debugging hard.

File: auth-server/src/controllers/public.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    remove(beaconId);
+    add(pubBeacon);
+
+    response.status(200).send(pubBeacon);
+  }
+);
+
+publicRouter.delete(
+  '/:beaconId',
+  (request: Request & DecodedToken, response: Response) => {
```
Comment (je-l) [inline auth-server/src/controllers/public.ts]: async/await here too

File: auth-server/src/middleware/requireLogin.ts [hunk 4]
```diff
@@ -32,7 +40,33 @@ const requireLogin = (req: Request, res: Response, next: () => void) => {
...
+
+  let decodedToken;
+  try {
+    decodedToken = jwt.verify(token, appConfig.JWT_SECRET) as any;
+  } catch (exception) {
+    return res.status(401).json({ error: 'invalid token' });
+  }
+
+  if (!decodedToken.beaconId || decodedToken.beaconId.length === 0) {
+    return res.status(401).json({ error: 'invalid token' });
```
Comment (je-l) [inline auth-server/src/middleware/requireLogin.ts]: Please give more informative error messages for the api user. Something which indicates why the error happened.

Comment (je-l) [reply auth-server/src/middleware/requireLogin.ts]: We have two similar errors with same explanation, it makes it hard to see quickly what is the reason for 401 error.

File: client/src/router.tsx (hunk 1 above)
Comment (je-l) [reply client/src/router.tsx]: It's probably related to the test runs having different environment, maybe API urls are wrong or something. You can use `it.skip` function to skip tests. I'd avoid commenting out code or adding try/catch blocks for just test runs.

Comment (je-l) [reply client/src/router.tsx]: It's probably related to differences in test environment, maybe API urls are wrong or something like that. You can skip tests with `it.skip`, i'd avoid commenting out code or adding try/catch blocks for just test runs.

**Pull request merged**